.config/emacs/config.org
#+TITLE: Mer Thich's Emacs Configuration
#+AUTHOR: Mer Thich (programming-programmer)

 * Very evil things follow. Beware! 👻

* Preface
 * This is copied from what_it_was.txt
 * Make sure init.el's path to config.org is there
 * use-package making sure is installed doesn't work either so install before

 * If there is an error with org, start emacs with "emacs -q" and install org
9.x. Then reload emacs

 * Install Noto Color Emoji or another emoji font to use emojis in Emacs
 * Also! Install Source Code Pro
 * Download ispell (via package manager) for flyspell
 * Also download java and languagetool.jar for langtool
 * Not that hard once you get to it. Java is getting easier each install 
* Variables
 * It's a bit much too have a seperate file for this
#+begin_src emacs-lisp
  (defvar mt/home (concat (getenv "HOME") "/") "My home directory.")
  (defvar mt/org-notes "/home/undefined_user/Notes" "Org Notes directory")

  (setq bookmark-default-file (concat mt/home ".cache/emacs/bookmarks"))
#+end_src

* System
** Garbage Collection
 * Make startup faster by reducing the frequency of garbage collection and then use a hook to measure Emacs startup time. The default is 800 kilobytes. Measured in bytes. Garbage collection off during initialization (focus all memory on initialize)
#+begin_src emacs-lisp
  (setq gc-cons-threshold (* 50 1000 1000) 
        gc-cons-percentage 0.6)
#+end_src

 * Reset garbage collection after initialization (return deprecated memory to stack when idle)
#+begin_src emacs-lisp
  (setq gcmh-idle-delay 'auto  ; default is 15s
        gcmh-auto-idle-delay-factor 10
        gcmh-high-cons-threshold (* 16 1024 1024))  ; 16mb

  (add-hook 'emacs-startup-hook
            (lambda ()
              (setq gc-cons-percentage 0.1))) ;; Default value for `gc-cons-percentage'

  (add-hook 'emacs-startup-hook
            (lambda ()
              (message "Emacs ready in %s with %d garbage collections."
                       (format "%.2f seconds"
                               (float-time
                                (time-subtract after-init-time before-init-time)))
                       gcs-done)))
#+end_src

** Package Management
*** Package Sources 
 * Where to gather package sources (MELPA and MELPA stable, and ELPA)
#+begin_src emacs-lisp
  (require 'package)
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("melpa-stable" . "https://stable.melpa.org/packages/")
                           ("elpa" . "https://elpa.gnu.org/packages/")))
#+end_src

*** Use-package
 * Use use-package for easier and more concise package management. yay!
#+begin_src emacs-lisp
  (require 'use-package)
  (setq use-package-always-ensure t)
#+end_src

** Defaults in Emacs
#+begin_src emacs-lisp
  (setq-default
   ad-redefinition-action 'accept                   ; Silence warnings for redefinition
   inhibit-splash-screen t                          ; No splash screen
   sentence-end-double-space nil                    ; Double space after a period!? Inhumane!
   cursor-in-non-selected-windows t                 ; Hide the cursor in inactive windows
   initial-scratch-message "Welcome to the Church of Emacs!!\n     ^\nʕ•́ᴥ•̀ʔっ"
   fill-column 100                                  ; Set width for automatic line breaks
   help-window-select t                             ; Focus new help windows when opened
   indent-tabs-mode t                               ; Prefer tabs over inferior spaces
   inhibit-startup-screen t                         ; Disable start-up screen
   load-prefer-newer t                              ; Prefer the newest version of a file
   mark-ring-max 128                                ; Maximum length of mark ring
   select-enable-clipboard t                        ; Merge system's and Emacs' clipboard
   tab-width 4                                      ; Set width for tabs
   vc-follow-symlinks t                             ; Always follow the symlinks
   view-read-only t                                 ; Always open read only files in view mode
   display-time-default-load-average nil
   display-line-numbers-type 'visual
   display-time-mail-string ""
   warning-minimum-level :emergency
   browse-url-browser-function 'browse-url-generic
   browse-url-generic-program "/usr/bin/qutebrowser")

  (fset 'yes-or-no-p 'y-or-n-p)                     ; Replace yes/no prompts with y/n
  (set-default-coding-systems 'utf-8)               ; Default to utf-8 encoding
  (show-paren-mode 1)                               ; Show parent paranthesis
  (global-visual-line-mode nil)                     ; Turn off that silly visual line mode
  (savehist-mode t)                                 ; Save history
  (menu-bar-mode -1)                                ; Remove that menubar pls :)
  (tool-bar-mode -1)                                ; Remove toolbar too
  (scroll-bar-mode -1)                              ; Disable visible scrollbar
  (tooltip-mode -1)                                 ; Disable tooltips
  (set-fringe-mode 10)                              ; Give some breathing room  
  (display-line-numbers-mode -1)                    ; No line numbers
  (display-time-mode 1)                             ; Display time in cwispy modeline

  (setq org-link-frame-setup
        '((vm . vm-visit-folder-other-frame)
          (vm-imap . vm-visit-imap-folder-other-frame)
          (gnus . org-gnus-no-new-news)
          (file . find-file)
          (wl . wl-other-frame)))
#+end_src

** Transparency
 * If this works, I'll cry from happiness
#+begin_src emacs-lisp
  (set-frame-parameter nil 'alpha-background 85)

  (add-to-list 'default-frame-alist '(alpha-background . 85))
#+end_src
 
** Version Control
#+begin_src emacs-lisp
  (use-package git-auto-commit-mode
    :config
    (setq
     gac-ask-for-summary-p nil
     gac-automatically-add-new-files-p t
     gac-automatically-push-p t
     gac-silent-message-p t))
#+end_src

** Scrolling improvements
#+begin_src emacs-lisp
  (setq mouse-wheel-scroll-amount '(1 ((shift) . 1))) ; one line at a time
  (setq mouse-wheel-progressive-speed nil)            ; don't accelerate scrolling
  (setq mouse-wheel-follow-mouse 't)                  ; scroll window under mouse
  (setq scroll-step 1)                                ; keyboard scroll one line at a time
#+end_src

** Backups/Lockfiles
#+begin_src emacs-lisp
  (setq create-lockfiles nil
        make-backup-files nil
        ;; But in case the user does enable it, some sensible defaults:
        version-control t     ; number each backup file
        backup-by-copying t   ; instead of renaming current file (clobbers links)
        delete-old-versions t ; clean up after itself
        kept-old-versions 5
        kept-new-versions 5
        backup-directory-alist (list (cons "." (concat user-emacs-directory "backup/"))))
#+end_src

** Load Lisp Files
 * SMALL BRAIN MOMENT!!! Apparently??? Loading stuff defined in these files
before loading the ACUTAL❗? files causes eror that gives mer BIG!!! headache
 * Point is, don't do dumb programming :)
#+begin_src emacs-lisp
  (load "/home/undefined_user/.config/emacs/lisp/funcs.el")
#+end_src

** Use System Trash
 * There has been too many times I've had a near heart attack from deleting.
Google Drive was there to prevent to much loss. However, in a move towards no
google™ we must *not* rely on such things
#+begin_src emacs-lisp
  (setq trash-directory (concat mt/home ".Trash"))
  (setq delete-by-moving-to-trash t)
#+end_src

** Start Maximized
 * Start fullscreen may change don't be surprised. Y'see
#+begin_src emacs-lisp
  (add-hook 'emacs-startup-hook 'toggle-frame-maximized)
#+end_src

* General
** Which-key
 * Shows which keys are avaliable. HAH! look at that pun. amazing
#+begin_src emacs-lisp
  (use-package which-key
    :init (which-key-mode)
    :defer 2
    :config
    (setq which-key-idle-delay 0.4))
#+end_src

** Super Save!
#+begin_src emacs-lisp
  (use-package super-save
    :diminish super-save-mode
    :defer 2
    :config
    (setq super-save-auto-save-when-idle t
          super-save-idle-duration 5 ;; after 5 seconds of not typing autosave
          super-save-triggers ;; Functions after which buffers are saved (switching window, for example)
          '(evil-window-next evil-window-prev balance-windows other-window next-buffer previous-buffer)
          super-save-max-buffer-size 10000000)
    (super-save-mode +1))

  ;; After super-save autosaves, wait __ seconds and then clear the buffer. I don't like
  ;; the save message just sitting in the echo area.
  (defun jib-clear-echo-area-timer ()
    (run-at-time "2 sec" nil (lambda () (message " "))))
  (advice-add 'super-save-command :after 'jib-clear-echo-area-timer)
#+end_src

** Registers
#+begin_src emacs-lisp
  (setq register-preview-delay 0) ;; Show registers ASAP

  (set-register ?c (cons 'file (concat user-emacs-directory "config.org")))

  (set-register ?1 (cons 'file (concat mt/org-notes "/refile.org")))
  (set-register ?2 (cons 'file (concat mt/org-notes "/personal.org")))
  (set-register ?3 (cons 'file (concat mt/org-notes "/work.org")))
  (set-register ?4 (cons 'file (concat mt/org-notes "/watchlist.org")))

  ;; Classes
  (set-register ?F (cons 'file (concat mt/org-notes "/FALL_2023/fall_2023.org")))
  (set-register ?d (cons 'file (concat mt/org-notes "/FALL_2023/Drivers_ED/driver_s_education.org")))
  (set-register ?e (cons 'file (concat mt/org-notes "/FALL_2023/ENGL_C102/engl_c102.org")))
  (set-register ?m (cons 'file (concat mt/org-notes "/FALL_2023/MATH_C151/math_c151.org")))
  (set-register ?p (cons 'file (concat mt/org-notes "/FALL_2023/POLS_C101/pols_c101.org")))
#+end_src

** Rainbow delimeters
 * See the nested parentheses (God send) but only in programming-mode
#+begin_src emacs-lisp
  (use-package rainbow-delimiters
    :defer t
    :hook (prog-mode . rainbow-delimiters-mode))
#+end_src

** Magit
 * Magit integration (git integration for emacs)
#+begin_src emacs-lisp
  (use-package magit
    :commands (magit-status)
    :custom
    (magit-display-buffer-function #'magit-display-buffer-same-window-except-diff-v1))
#+end_src

** OpenWith
 * THE ANSWER I NEEDED!! Basically open files with a certain extension with an
external program. For more extensions see David's config.daviwil
 * Maybe add an image viewer later??
#+begin_src emacs-lisp
  (use-package openwith
    :defer t
    :config
    (setq openwith-associations
          (list
           (list (openwith-make-extension-regexp
                  '("mpg" "mpeg" "mp3" "mp4"
                    "avi" "wmv" "wav" "mov" "flv"
                    "ogm" "ogg" "mkv"))
                 "mpv"
                 '(file))
           (list (openwith-make-extension-regexp
                  '("pdf"))
                 "zathura"
                 '(file))))
    (openwith-mode))
#+end_src

* Aesthetics
** Font
 * NOTE: On a new system, you will have to install Source Code Pro on the system
#+begin_src emacs-lisp
  (set-face-attribute 'default t :weight 'medium :height 100 :font "Source Code Pro")
#+end_src

 * Font for org-mode
#+begin_src emacs-lisp
  (custom-theme-set-faces
   'user
   '(variable-pitch ((t (:family "Source Code Pro" :weight medium))))
   '(fixed-pitch ((t ( :family "Source Code Pro" :height 200)))))
#+end_src 

** Theme
#+begin_src emacs-lisp
  (use-package doom-themes
    :config
    (load-theme 'doom-tomorrow-night t))
#+end_src

** Writeroom
#+begin_src emacs-lisp
  (use-package visual-fill-column
    :defer t
    :config
    (setq visual-fill-column-center-text t)
    (setq visual-fill-column-width 80)
    (setq visual-fill-column-center-text t))

  (use-package writeroom-mode
    :defer t
    :config
    (setq writeroom-maximize-window nil
          writeroom-mode-line nil
          writeroom-global-effects nil ;; No need to have Writeroom do any of that silly stuff
          writeroom-extra-line-spacing 3) 
    (setq writeroom-width visual-fill-column-width)
    )
#+end_src

** Modeline
*** DOOM Modeline
 * With the defined segements, there is no need for the diminish package and it
is _exactly_ how I want :)
#+begin_src emacs-lisp
  (use-package doom-modeline
    :config
    (doom-modeline-mode)
    (setq doom-modeline-project-detection 'file-name
          doom-modeline-highlight-modified-buffer-name t
          ;; Icons ---
          doom-modeline-modal t
          doom-modeline-icon nil
          doom-modeline-buffer-state-icon t
          doom-modeline-buffer-modification-icon nil
          doom-modeline-unicode-fallback nil
          doom-modeline-bar-width 3))

  (doom-modeline-def-modeline 'minimal
    '(bar modals buffer-info buffer-position)
    '(time major-mode))
  (doom-modeline-set-modeline 'minimal t)
#+end_src

*** Hide Modeline
 * Honestly, this is a blessing. Hide the modeline. This is mapped to a
keybinding (evil mode of course 👻)
#+begin_src emacs-lisp
  (use-package hide-mode-line
    :defer t
    :commands (hide-mode-line-mode))
#+end_src

*** Nyan Mode
 * Nyan mode baby!
#+begin_src emacs-lisp
  (use-package nyan-mode 
    :init (nyan-mode)
    :config
    (setq nyan-wavy-trail nil))
#+end_src

* Keybindings
 * All of this thanks to general.el 🎉🎉🎉🎉🎉 YAYAYAY!!
#+begin_src emacs-lisp
  (use-package general)
#+end_src

** Evil Mode
#+begin_src emacs-lisp
   (use-package evil
     :init
     (setq evil-want-keybinding nil) ;; load Evil keybindings in other modes
     (setq evil-want-fine-undo t)
     (setq evil-want-Y-yank-to-eol t)
     (setq evil-mode-line-format nil)
     (setq evil-disable-insert-state-bindings t)

     :config
     (define-key evil-motion-state-map "/" 'swiper)
     (evil-global-set-key 'motion "j" 'evil-next-visual-line)
     (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
     (define-key evil-normal-state-map (kbd "C-u") 'evil-scroll-up)
     (global-set-key [remap evil-quit] 'kill-buffer-and-window)

     (setq evil-emacs-state-cursor    '("#649bce" box))
     (setq evil-normal-state-cursor   '("#d9a871" box))
     (setq evil-operator-state-cursor '("#ebcb8b" hollow))
     (setq evil-visual-state-cursor   '("#677691" box))
     (setq evil-insert-state-cursor   '("#eb998b" box))
     (setq evil-replace-state-cursor  '("#eb998b" hbar))
     (setq evil-motion-state-cursor   '("#ad8beb" box))

     (evil-set-undo-system 'undo-fu)
     (evil-mode 1))

   (use-package evil-collection
     :after evil
     :config
     (evil-collection-init))

#+end_src

*** Undo/Redo Functionality (with undo-tree)
#+begin_src emacs-lisp
  (use-package undo-fu)
#+end_src

** SPC Leader Key
#+begin_src emacs-lisp
  (general-define-key
   :states '(normal motion visual)
   :keymaps 'override
   :prefix "SPC"
#+end_src

*** Top level functions
#+begin_src emacs-lisp
  "." '(counsel-find-file :which-key "find file")
  "r" '(counsel-recentf :which-key "recent files")
  "TAB" '(switch-to-prev-buffer :which-key "previous buffer")
  "SPC" '(counsel-M-x :which-key "M-x")
  "RET" '(counsel-bookmark :which-key "bookmarks")
  "," '(counsel-switch-buffer :which-key "switch buffers")
  "c" '(org-capture :which-key "org-capture")
  "z" '(repeat :which-key "repeat")
  "j" '(jump-to-register :which-key "jump to register")
#+end_src

*** Applications
#+begin_src emacs-lisp
  "o" '(nil :which-key "open")

  "oa" '(org-agenda :which-key "org-agenda")
  "oc" '(calendar :which-key "calendar")
  "oC" '(calc :which-key "calc")
  "o-" '(dired :which-key "dired")
  "od" '(dired-jump :which-key "dired jump")
  "of" '(make-frame :which-key "new frame")
  "oF" '(select-frame-by-name :which-key "select frame")
#+end_src

*** Buffers
#+begin_src emacs-lisp
  "b" '(nil :which-key "buffer")

  "bb" '(counsel-switch-buffer :which-key "switch buffers")
  "bd" '(evil-delete-buffer :which-key "delete buffer")
  "bn" '(evil-next-buffer :which-key "next buffer")
  "bp" '(previous-buffer :which-key "previous buffer")
  "bm" '(bookmark-save :which-key "set bookmark")
  "bM" '(bookmark-delete  :which-key "delete bookmark")
  "bN" '(evil-buffer-new :which-key "new empty buffer")
#+end_src

*** Files
#+begin_src emacs-lisp
  "f" '(nil :which-key "files")

  "fb" '(counsel-bookmark :which-key "bookmarks")
  "fC" '(copy-file :which-key "copy this file")
  "fD" '(delete-file :which-key "delete this file")
  "ff" '(counsel-find-file :which-key "find file")
  "fr" '(counsel-recentf :which-key "recent files")
  "fR" '(rename-file :which-key "rename/move file")
  "fs" '(save-buffer :which-key "save buffer")
  "fS" '(evil-write-all :which-key "save all buffers")
  "fl" '(org-babel-load-file :which-key "org-babel load file")
#+end_src

*** Quit/Session
#+begin_src emacs-lisp
  "q" '(nil :which-key "quit/session")

  "qd" '(nil :which-key "restart emacs server")
  "qf" '(delete-frame :which-key "delete frame")
  "qF" '(delete-other-frames :which-key "delete other frames")
  "ql" '(recover-session :which-key "restore last session")
#+end_src

*** Notes
#+begin_src emacs-lisp
  "n" '(nil :which-key "notes")

  "nc" '(org-clock-in-last :which-key "toggle last org-clock")
  "nC" '(org-clock-out :which-key "clock out of current org-clock")
  "n C-c" '(org-clock-cancel :which-key "cancel current org-clock")
  "no" '(org-clock-goto :which-key "active org-clock")
  "np" '(org-pomodoro :which-key "start pomodoro")
  "nd" '(deft :which-key "open deft")

  "nr" '(nil :which-key "org-roam")
  "nrl" '(org-roam-buffer-toggle :which-key "toggle buffer")
  "nrf" '(org-roam-node-find :which-key "find node")
  "nrc" '(org-roam-capture :which-key "capture node")
  "nri" '(org-roam-node-insert :which-key "insert node")
  "nrI" '(org-roam-node-insert-immediate :which-key "immediately insert node")
#+end_src

*** Help/Emacs
#+begin_src emacs-lisp
  "h" '(nil :which-key "help/emacs")

  "h RET" '(info-emacs-manual :which-key "info-emacs-manual")
  "h'" '(describe-char :which-key "describe-char")
  "h." '(display-local-help :which-key "desplay-local-help")
  "h?" '(help-for-help :which-key "help-for-help")
  "hC" '(describe-coding-system :which-key "describe-coding-system")
  "he" '(view-echo-area-messages :which-key "view-echo-area-messages")
  "hf" '(counsel-describe-function :which-key "describe function")
  "hF" '(counsel-describe-face :which-key "describe-face")
  "hg" '(describe-gnu-project :which-key "describe-gnu-project")
  "hi" '(info :which-key "info")
  "hv" '(describe-variable :which-key "describe variable")
  "hI" '(describe-input-method :which-key "describe-input-method")
  "hk" '(describe-key :which-key "describe-key")
  "hm" '(describe-mode :which-key "describe-mode")

  "hp" '(nil :which-key "packages")
  "hpr" '(package-refresh-contents :which-key "refresh packages")
  "hpi" '(package-install :which-key "install packages")
  "hpd" '(package-delete :which-key "delete package")
  "hpD" '(describe-package :which-key "describe package")
#+end_src

*** Insert
#+begin_src emacs-lisp
  "i" '(nil :which-key "insert")

  "ie" '(emoji-search :which-key "emoji")
  "iu" '(counsel-unicode-char :which-key "unicode")
  "iy" '(counsel-yank-pop :which-key "from clipboard")
  "ij" '(point-to-register :which-key "insert register")
  "ii" '(org-id-get-create :which-key "insert org-id")
  "il" '(org-insert-link :which-key "insert link")
  "is" '(nil :which-key "insert stamp")
  "iss" '((lambda () (interactive) (call-interactively (org-time-stamp-inactive))) :which-key "org-time-stamp-inactive")
  "isS" '((lambda () (interactive) (call-interactively (org-time-stamp nil))) :which-key "org-time-stamp")

#+end_src

*** Toggles/Visuals
#+begin_src emacs-lisp
  "t" '(nil :which-key "toggles")
  "ty" '(counsel-load-theme 'doom-spacegray :which-key "lighter theme")
  "tt" '(toggle-truncate-lines :which-key "truncate lines")
  "tv" '(visual-line-mode :which-key "visual line mode")
  "tn" '(display-line-numbers-mode :which-key "display line numbers")
  "tR" '(read-only-mode :which-key "read only mode")
  "tw" '(writeroom-mode :which-key "writeroom mode")
  "tm" '(hide-mode-line-mode :which-key "hide modeline mode")
#+end_src

*** End SPC prefix block
#+begin_src emacs-lisp
  )
#+end_src

** All-mode keybindings
#+begin_src emacs-lisp
  (general-def
    :keymaps 'override
    )
#+end_src

** Insert Mode Keymaps
#+begin_src emacs-lisp
  ;; Insert keymaps
  (general-def
    :states '(insert)

    "C-c" 'evil-normal-state

    ;; Emacs ---
    "C-x C-m" 'counsel-M-x

    ;; Utility ---
    "C-s" 'swiper

    "M-k" 'org-metaup
    "M-j" 'org-metadown
    "S-M-l" 'org-shiftmetaright
    "S-M-h" 'org-shiftmetaleft
    "M-l" 'org-metaright
    "M-h" 'org-metaleft

    "<tab>" 'org-cycle
    "TAB" 'org-cycle

    "<S-up>" 'org-increase-number-at-point
    "<S-down>" 'org-decrease-number-at-point

    )
#+end_src

** Normal Mode Keymaps
#+begin_src emacs-lisp
  ;; Insert keymaps
  (general-def
    :states '(normal)
    "$" 'evil-end-of-visual-line
    "0" 'evil-beginning-of-visual-line

    "gf" 'org-open-at-point

    ;;Windows
    "C-w m" '(jib/toggle-maximize-buffer :which-key "maximize window")

    "<S-up>" 'org-increase-number-at-point
    "<S-down>" 'org-decrease-number-at-point

    )
#+end_src

** Org mode Keybindings
#+begin_src emacs-lisp
  (general-define-key
   :prefix "SPC m"
   :states '(normal visual motion)
   :keymaps '(org-mode-map)
   "" nil
   "A" '(org-archive-subtree-default :which-key "org-archive")
   "g" '(counsel-org-goto :which-key "goto heading")
   ":" '(counsel-org-tag :which-key "set tags")
   "P" '(org-set-property :which-key "set property")
   "E" '(org-export-dispatch :which-key "export org")
   "C-e" '(export-org-email :which-key "export org email")
   "e" '(org-set-effort :which-key "set effort")
   "." '(org-toggle-narrow-to-subtree :which-key "toggle narrow to subtree")
  
   "s" '(org-schedule :which-key "schedule")
   "S" '(jib/org-schedule-tomorrow :which-key "schedule tmrw")
   "d" '(org-deadline :which-key "deadline")

   "t" '(org-todo :which-key "toggle TODO state")
   "C" '(org-toggle-checkbox :which-key "toggle checkbox")

   "1" '(org-toggle-link-display :which-key "toggle link display")
   "2" '(org-toggle-inline-images :which-key "toggle images")
   "6" '(org-sort :which-key "sort")

   "T" '(nil :which-key "org-transclusion")
   "Ta" '(org-transclusion-add :which-key "add org-transclusion")
   "Tt" '(org-transclusion-mode :which-key "org-transclusion mode")
   )
#+end_src

** Org Agenda Keybindings
#+begin_src emacs-lisp
(general-define-key
 :prefix ","
 :states '(normal motion visual)
 :keymaps '(org-agenda-mode-map)
 "" nil
 "c" '(org-capture :which-key "org-capture")
 "d" '(org-agenda-deadline :which-key "deadline")
 "s" '(org-agenda-schedule :which-key "schedule") 
 "t" '(org-agenda-set-tags :which-key "set tags")

 "c" '(nil :which-key "clocking")
 "ci" '(org-agenda-clock-in :which-key "clock in")
 "co" '(org-agenda-clock-out :which-key "clock out")
 "cj" '(org-clock-goto :which-key "jump to clock")
 )

(evil-define-key 'motion org-agenda-mode-map
  (kbd "f") 'org-agenda-later
  (kbd "b") 'org-agenda-earlier)
#+end_src

** Dired
#+begin_src emacs-lisp
  (evil-define-key 'normal dired-mode-map
    (kbd "M-RET") 'dired-display-file
    (kbd "h") 'dired-up-directory
    (kbd "l") 'dired-find-alternate-file)
#+end_src

** Duplicate line
#+begin_src emacs-lisp
    (defun duplicate-line()
      (interactive)
      (move-beginning-of-line 1)
      (kill-line)
      (yank)
      (open-line 1)
      (next-line 1)
      (yank))
#+end_src

* Ivy
 * Ivy is an excellent completion framework for Emacs. It provides a minimal yet powerful selection menu that appears when you open files, switch buffers, and for many other tasks in Emacs.
#+begin_src emacs-lisp
  (use-package ivy
    :diminish ivy-mode
    :config
    (setq ivy-extra-directories nil) ;; Hides . and .. directories
    (setq ivy-initial-inputs-alist nil) ;; Removes the ^ in ivy searches
    :config
    (setq ivy-use-virtual-buffers t)
    (setq ivy-wrap t)
    (setq ivy-count-format "(%d/%d) ")
    (setq enable-recursive-minibuffers t)

    (ivy-mode 1))  
#+end_src

** Ivy Rich
 * ivy-rich adds extra columns to a few of the Counsel commands to provide more information about each item.
 #+begin_src emacs-lisp
   (use-package ivy-rich
    :init
    (setq ivy-rich-path-style 'abbrev)
    (ivy-rich-mode 1))
 #+end_src

** Counsel
 * Counsel is a customized set of commands to replace `find-file` with
   `counsel-find-file`, etc which provide useful commands for each of the default completion commands.
 * Allows for selection. Usually you wouldn't be able to browse through the
   options in the minibuffer but w/ counsel you can
#+begin_src emacs-lisp
  (use-package counsel
    :bind (("C-x b" . 'counsel-switch-buffer))
    :config
    (counsel-mode 1))
#+end_src

* Org Mode
** Org Packages
*** Org-indent Face
 * Make sure org-indent face is avaliable
 * Makes it so that the headers indent
#+begin_src emacs-lisp
    (require 'org-indent)
#+end_src

*** Org-habit
#+begin_src emacs-lisp
  (require 'org-habit)
  (setq org-habit-graph-column 60)
  (add-to-list 'org-modules 'org-habit)
#+end_src

*** Org-roam
#+begin_src emacs-lisp
  (use-package org-roam
    :after org
    :config
    (org-roam-setup)
    (setq org-roam-completion-system 'ivy-mode)

    :custom
    (org-roam-directory (concat mt/org-notes))
    (org-roam-completion-everywhere t)
#+end_src

**** Org Roam Capture Templates
#+begin_src emacs-lisp
   (org-roam-capture-templates '(
#+end_src

***** Default Template
#+begin_src emacs-lisp
  ("d" "default" plain
   "%?"
   :if-new (file+head "${slug}.org" "#+title: ${title}\n")
   :unnarrowed t)
#+end_src

***** School Note Template
#+begin_src emacs-lisp

.config/qutebrowser/config.py
# Web Browsers:
c.url.searchengines = {
    "DEFAULT": "https://search.brave.com/search?q={}", 
    "g": "https://www.google.com/search?q={}", 
    "yt": "https://www.youtube.com/results?search_query={}",
    "myt": "https://music.youtube.com/search?q={}",
    "dict": "https://www.merriam-webster.com/dictionary/{}",
    "syn": "https://www.merriam-webster.com/thesaurus/{}",
    "nix": "https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query={}"
}

# Start and Default pages
c.url.default_page = "about:blank"
c.url.start_pages = "about:blank"

# Theme and Dark Mode
config.source("nord-qutebrowser.py")
c.colors.webpage.preferred_color_scheme = 'dark'
c.colors.webpage.darkmode.enabled = True
c.colors.webpage.darkmode.algorithm = "lightness-cielab"

c.input.insert_mode.auto_leave = True
c.input.insert_mode.auto_load = False
c.input.insert_mode.leave_on_load = True
c.hints.chars = "dhtns"
c.hints.uppercase = True
c.tabs.last_close = "startpage"
c.tabs.mode_on_change = "normal"
c.tabs.title.format = "{index}: {current_title}{private}"
c.window.hide_decoration = True
c.bindings.key_mappings = {"<Ctrl-c>": "<Escape>"}
c.content.autoplay = False
c.window.transparent = True
c.statusbar.widgets = ['keypress', 'history', 'tabs']
c.statusbar.show = "always"
c.statusbar.padding = {'bottom': 0, 'left': 0, 'right': 0, 'top': 0}
c.spellcheck.languages = ['en-US']
c.content.tls.certificate_errors = "load-insecurely"
c.scrolling.bar = "never"
c.completion.quick = True
c.content.pdfjs = True
c.content.mute = True
c.content.notifications.presenter = "messages"
c.messages.timeout = 1000

# General Bindings
config.bind(";M", "hint --rapid links spawn umpv {hint-url}")
config.bind(",m", "hint links spawn -v --detach umpv '{hint-url}'") 
config.bind(",M", "spawn -v --detach umpv '{url}'") 
config.bind(",s", "hint links spawn -v --detach streamlink --player mpv '{hint-url}' 720p") 
config.bind(",S", "spawn -v --detach streamlink --player mpv '{url}' 720p") 
config.bind("C", "clear-messages")
config.bind(",F", "hint links spawn -v firefox {hint-url}")
config.bind(",a", "adblock-update")
config.bind("<Ctrl+h>", "back --tab")
config.bind("<Ctrl+l>", "forward --tab")

config.bind("f", "hint all")
config.bind("gD", "set-cmd-text -s :tab-give")
config.bind("M", "set-cmd-text -s :bookmark-add {url} """)
config.bind(";f", "hint --rapid all current")
config.unbind("+")
config.unbind("-")
config.unbind("=")
config.bind("<Ctrl-0>", "zoom")
config.bind("<Ctrl-=>", "zoom-in")
config.bind("<Ctrl-->", "zoom-out")
config.unbind("d")
config.bind("X", "tab-close")
config.bind("K", "tab-next")
config.bind("J", "tab-prev")
config.bind("D", "tab-close --force")

# Insert Bindings
config.bind("<Ctrl-f>", "fake-key <Right>", mode="insert")
config.bind("<Ctrl-b>", "fake-key <Left>", mode="insert")
config.bind("<Ctrl-a>", "fake-key <Home>", mode="insert")
config.bind("<Ctrl-e>", "fake-key <End>", mode="insert")
config.bind("<Ctrl-n>", "fake-key <Down>", mode="insert")
config.bind("<Ctrl-p>", "fake-key <Up>", mode="insert")
config.bind("<Alt-f>", "fake-key <Ctrl-Right>", mode="insert")
config.bind("<Alt-b>", "fake-key <Ctrl-Left>", mode="insert")
config.bind("<Ctrl-d>", "fake-key <Delete>", mode="insert")
config.bind("<Alt-d>", "fake-key <Ctrl-Delete>", mode="insert")
config.bind("<Alt-backspace>", "fake-key <Ctrl-Backspace>", mode="insert")
config.bind("<Ctrl-w>", "fake-key <Ctrl-backspace>", mode="insert")
config.bind("<Ctrl-y>", "insert-text {clipboard}", mode="insert")
config.bind("<Escape>", "mode-leave", mode="insert")

# Command Bindings
config.bind("<Ctrl-p>", "completion-item-focus prev", mode="command")
config.bind("<Ctrl-n>", "completion-item-focus next", mode="command")
config.bind("<Alt-b>", "rl-backward-word", mode="command")
config.bind("<Alt-d>", "rl-kill-word", mode="command")
config.bind("<Alt-f>", "rl-forward-word", mode="command")
config.bind("<Ctrl-a>", "rl-beginning-of-line", mode="command")
config.bind("<Ctrl-b>", "rl-backward-char", mode="command")
config.bind("<Ctrl-c>", "completion-item-yank", mode="command")
config.bind("<Ctrl-d>", "completion-item-del", mode="command")
config.bind("<Ctrl-e>", "rl-end-of-line", mode="command")
config.bind("<Ctrl-f>", "rl-forward-char", mode="command")
config.bind("<Ctrl-h>", "rl-backward-delete-char", mode="command")
config.bind("<Ctrl-k>", "rl-kill-line", mode="command")
config.bind("<Ctrl-c>", "mode-leave", mode="command")

# Caret mode
config.bind("$", "move-to-end-of-line", mode="caret")
config.bind("0", "move-to-start-of-line", mode="caret")
config.bind("<Escape>", "mode-leave", mode="caret")
config.bind("G", "move-to-end-of-document", mode="caret")
config.bind("H", "scroll left", mode="caret")
config.bind("J", "scroll down", mode="caret")
config.bind("K", "scroll up", mode="caret")
config.bind("L", "scroll right", mode="caret")
config.bind("V", "selection-toggle --line", mode="caret")
config.bind("Y", "yank selection -s", mode="caret")
config.bind("[", "move-to-start-of-prev-block", mode="caret")
config.bind("]", "move-to-start-of-next-block", mode="caret")
config.bind("b", "move-to-prev-word", mode="caret")
config.bind("c", "mode-enter normal", mode="caret")
config.bind("e", "move-to-end-of-word", mode="caret")
config.bind("gg", "move-to-start-of-document", mode="caret")
config.bind("h", "move-to-prev-char", mode="caret")
config.bind("j", "move-to-next-line", mode="caret")
config.bind("k", "move-to-prev-line", mode="caret")
config.bind("l", "move-to-next-char", mode="caret")
config.bind("o", "selection-reverse", mode="caret")
config.bind("v", "selection-toggle", mode="caret")
config.bind("w", "move-to-next-word", mode="caret")
config.bind("y", "yank selection", mode="caret")
config.bind("{", "move-to-end-of-prev-block", mode="caret")
config.bind("}", "move-to-end-of-next-block", mode="caret")

# Bindings for yesno mode
config.bind("<Escape>", "mode-leave", mode="yesno")
config.bind("<Return>", "prompt-accept", mode="yesno")
config.bind("N", "prompt-accept --save no", mode="yesno")
config.bind("Y", "prompt-accept --save yes", mode="yesno")
config.bind("n", "prompt-accept no", mode="yesno")
config.bind("y", "prompt-accept yes", mode="yesno")

config.bind("<Escape>", "mode-leave", mode="hint")
config.bind('<Ctrl-g>', 'mode-leave', mode='passthrough')

# Load Autoconfig
config.load_autoconfig(True)
